models/credmark/protocols/lending/aave/aave_v2.py

```diff
@@ -86,96 +123,121 @@ def run(self, input) -> Portfolio:
                 version="1.0",
                 display_name="Aave V2 token liquidity",
                 description="Aave V2 token liquidity at a given block number",
                 input=Token,
                 output=AaveDebtInfo)
 class AaveV2GetTokenAsset(Model):
-
     def run(self, input: Token) -> AaveDebtInfo:
-
-        contract = Contract(
-            # AAVE Lending Pool V2
-            address=Address(AAVE_LENDING_POOL_V2).checksum,
-            abi=AAVE_V2_TOKEN_CONTRACT_ABI
-        )
-
-        reservesData = contract.functions.getReserveData(input.address).call()
-
-        aToken = Token(address=reservesData[7], abi=ERC_20_TOKEN_CONTRACT_ABI)
-        stableDebtToken = Token(address=reservesData[8], abi=ERC_20_TOKEN_CONTRACT_ABI)
-        variableDebtToken = Token(address=reservesData[9], abi=ERC_20_TOKEN_CONTRACT_ABI)
+        aave_lending_pool = self.context.run_model('aave-v2.get-lending-pool',
+                                                   input=EmptyInput(),
+                                                   return_type=Contract)
+
+        reservesData = aave_lending_pool.functions.getReserveData(input.address).call()
+
+        # reservesData
+        # | Name | Type | Description |
+        # 0. configuration | uint256 | bitmark
+        # 1. liquidityIndex | uint128 | liquidity index in ray (1e27) |
+        # / 1e27 ~= aToken.functions.totalSupply().call() /
+        #           aToken.functions.scaledTotalSupply().call()
+        # 2. variableBorrowIndex | uint128 | variable borrow index in ray |
+        # / 1e27 ~= variableDebtToken.functions.totalSupply().call() /
+        #           variableDebtToken.functions.scaledTotalSupply().call()
+
+        # 3. currentLiquidityRate | uint128 | current supply / liquidity / deposit rate in ray |
+        # 4. currentVariableBorrowRate | uint128 | current variable borrow rate in ray |
+        # 5. currentStableBorrowRate | uint128 | current stable borrow rate in ray |
+        # 6. lastUpdateTimestamp | uint40 | timestamp of when reserve data was last updated |
+
+        # 7. aTokenAddress | address | address of associated aToken (tokenised deposit) |
+        # 8. stableDebtTokenAddress | address | address of associated stable debt token |
+        # 9. variableDebtTokenAddress | address | address of associated variable debt token |
+        # 10. interestRateStrategyAddress | address | address of interest rate strategy
+        # 11. id | uint8 | the position in the list of active reserves |
+
+        aToken = Token(address=reservesData[7])
+        stableDebtToken = Token(address=reservesData[8], abi=AAVE_STABLEDEBT_ABI)
+        variableDebtToken = Token(address=reservesData[9])
         interestRateStrategyContract = Contract(address=reservesData[10])
 
-        totalStableDebt = stableDebtToken.total_supply
-        totalVariableDebt = variableDebtToken.total_supply
+        currentLiquidityRate = reservesData[3] / 1e27
```

Comment:
Is there a generalized `Reserves` object that we can make and put in the framework?

Reply:
Common pattern starts to emerge. I will try adding one.